personalities/sctfe/handlers.go

```diff
@@ -326,21 +331,36 @@ func addChainInternal(ctx context.Context, li *logInfo, w http.ResponseWriter, r
 		return http.StatusBadRequest, fmt.Errorf("failed to build MerkleTreeLeaf: %s", err)
 	}
 
-	// TODO(phboneff): refactor entryFromChain to avoid recomputing hashes in AddIssuerChain
-	if len(chain) > 1 {
+	klog.V(2).Infof("%s: %s => storage.GetCertIndex", li.LogOrigin, method)
+	idx, ok, err := li.storage.GetCertIndex(ctx, chain[0])
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("couldn't deduplicate the request: %s", err)
+	}
+
+	if ok {
+		klog.V(3).Infof("%s: %s - found duplicate entry at index %d", li.LogOrigin, method, idx)
+	} else {
 		if err := li.storage.AddIssuerChain(ctx, chain[1:]); err != nil {
 			return http.StatusInternalServerError, fmt.Errorf("failed to store issuer chain: %s", err)
 		}
-	}
 
-	klog.V(2).Infof("%s: %s => storage.Add", li.LogOrigin, method)
-	idx, err := li.storage.Add(ctx, entry)
-	if err != nil {
-		if errors.Is(err, tessera.ErrPushback) {
-			w.Header().Add("Retry-After", "1")
-			return http.StatusServiceUnavailable, fmt.Errorf("Tessera sequencer pushed back: %v", err)
+		klog.V(2).Infof("%s: %s => storage.Add", li.LogOrigin, method)
+		idx, err = li.storage.Add(ctx, entry)
+		if err != nil {
+			if errors.Is(err, tessera.ErrPushback) {
+				w.Header().Add("Retry-After", "1")
+				return http.StatusServiceUnavailable, fmt.Errorf("Tessera sequencer pushed back: %v", err)
+			}
+			return http.StatusInternalServerError, fmt.Errorf("couldn't store the leaf: %v", err)
 		}
-		return http.StatusInternalServerError, fmt.Errorf("couldn't store the leaf: %v", err)
+		go func() {
+			klog.V(2).Infof("%s: %s => storage.AddCertIndex", li.LogOrigin, method)
+			err := li.storage.AddCertIndex(ctx, chain[0], idx)
```

Comment:
IIUC we'll set this here, and then if we happen to be using a "local" dedup storage we'll try to set it again when the tracker finds it integrated into the log?

Might be worth a comment about that.

Reply:
That's correct, done.